### PR TITLE
Add support for wide angle camera in sensors system

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

depends on:
* #1183
* https://github.com/ignitionrobotics/ign-sensors/pull/171
* https://github.com/ignitionrobotics/ign-rendering/pull/490
* https://github.com/ignitionrobotics/sdformat/pull/744

## Summary

supports parsing and loading wide angle camera sensors

## Test it

you can test with the `wide_angle_camera.sdf` example world:

```
ign gazebo -v 4 wide_angle_camera.sdf
```

wide_angle_camera_ign_gazebo

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Review by ahcorde — COMMENTED** [review]
is this targeting the right branch ?

**Comment by iche033** [comment]
> is this targeting the right branch ?

yeah, this targets garden and requires the version bumps for gui, sensors and rendering done in #1183

**Comment by WilliamLewww** [comment]
All components (ign-rendering, ign-sensors, ign-gazebo) are working as expected. The test world is also working as expected.

**Review by WilliamLewww — APPROVED** [review]

**Comment by scpeters** [comment]
I tried adding a `dependencies.yaml` file in 77a0f602a6f4ba208facc3bf1c68513af938f15b in branch `wide_angle_camera_vcs` to build ign-rendering and ign-sensors from source to see if that fixes the GitHub actions workflow CI

[Ubuntu CI](https://github.com/ignitionrobotics/ign-gazebo/actions/workflows/ci.yml)

**Comment by scpeters** [comment]
> I tried adding a `dependencies.yaml` file in [77a0f60](https://github.com/ignitionrobotics/ign-gazebo/commit/77a0f602a6f4ba208facc3bf1c68513af938f15b) in branch `wide_angle_camera_vcs` to build ign-rendering and ign-sensors from source to see if that fixes the GitHub actions workflow CI
> 
> [Ubuntu CI](https://github.com/ignitionrobotics/ign-gazebo/actions/workflows/ci.yml)

it built ign-sensors7 successfully and ign-gazebo7 configured correctly. It later failed with an unrelated linking error due to branches being out of date with `main`

File: examples/worlds/multi_lrauv_race.sdf [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" ?>
 
 <!-- Run demo with ign-gazebo examples/standalone/multi_lrauv_race
-  (example is provided in ign-gazebo6+). -->
+  (example is provided in ign-gazebo7+). -->
```
Comment (chapulina) [inline examples/worlds/multi_lrauv_race.sdf:4]: I think this can be reverted.

File: src/gui/plugins/spawn/Spawn.cc [hunk 2]
```diff
@@ -190,7 +190,7 @@ void Spawn::LoadConfig(const tinyxml2::XMLElement *)
 }
 
 
-// TODO(ahcorde): Replace this when this function is on ign-rendering6
+// TODO(ahcorde): Replace this when this function is on ign-rendering7
```
Comment (chapulina) [inline src/gui/plugins/spawn/Spawn.cc:193]: Another change that I think can be reverted. The function is already on `ign-rendering6`.

**Review by chapulina — COMMENTED** [review]
We need to add the wide angle camera package to https://github.com/ignition-release/ign-sensors7-release for the nightlies to work.

**Comment by scpeters** [comment]
> We need to add the wide angle camera package to https://github.com/ignition-release/ign-sensors7-release for the nightlies to work.

indeed, that's why this build is unstable:

[Build Status](https://build.osrfoundation.org/view/ign-garden/job/ign-sensors7-debbuilder/58/) https://build.osrfoundation.org/view/ign-garden/job/ign-sensors7-debbuilder/58/

I'll fix it

**Comment by scpeters** [comment]
> > We need to add the wide angle camera package to https://github.com/ignition-release/ign-sensors7-release for the nightlies to work.
> 
> indeed, that's why this build is unstable:
> 
> [Build Status](https://build.osrfoundation.org/view/ign-garden/job/ign-sensors7-debbuilder/58/) https://build.osrfoundation.org/view/ign-garden/job/ign-sensors7-debbuilder/58/
> 
> I'll fix it

https://github.com/ignition-release/ign-sensors7-release/pull/4

**Comment by scpeters** [comment]
CI is looking better

File: src/gui/plugins/spawn/Spawn.cc (hunk 2 above)
Comment (iche033) [reply src/gui/plugins/spawn/Spawn.cc]: reverted. c0657e52630f5020644d814ca7f2a391d44d9f49

File: examples/worlds/multi_lrauv_race.sdf (hunk 1 above)
Comment (iche033) [reply examples/worlds/multi_lrauv_race.sdf]: reverted. c0657e52630f5020644d814ca7f2a391d44d9f49

**Comment by scpeters** [comment]
does this need an integration test before merging? I can find someone to add that to this branch if so

**Comment by iche033** [comment]
one simple integration test I can think of is to load the wide angle camera world file and make sure data are published to the topic. ign-sensors and ign-rendering already have integration tests to check the content of the published images so we probably don't need to duplicate that here.

yeah it would be nice to get some help on adding the test.

**Comment by scpeters** [comment]
the new test failed in https://github.com/ignitionrobotics/ign-gazebo/pull/1215/commits/5612cb6b6ca85670c7e99d58e87bab92bcd02a27

* https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/7838/testReport/junit/(root)/INTEGRATION_wide_angle_camera/test_ran/

File: test/integration/wide_angle_camera.cc [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+/////////////////////////////////////////////////
+void imageCb(const msgs::Image &_msg)
+{
+  mutex.lock();
+  unsigned int imageSamples = _msg.width() * _msg.height();
+  unsigned int imageBufferSize = imageSamples * sizeof(float);
+
+  if (!imageBuffer)
+    imageBuffer = new float[imageSamples];
+  memcpy(imageBuffer, _msg.data().c_str(), imageBufferSize);
```
Comment (scpeters) [inline test/integration/wide_angle_camera.cc:56]: I got a backtrace indicating that the seg-fault occurs at this line

**Comment by scpeters** [comment]
looks like the test is fixed in https://github.com/ignitionrobotics/ign-gazebo/pull/1215/commits/d6a39d0a623091fe7ae46f62d6d2abeb7954d28b, though there are some unrelated cmake warnings

**Comment by WilliamLewww** [comment]
Still running into a segfault from using an incorrect pixel format. Currently trying to fix the error so it runs smoothly like the depth camera integration test.

**Comment by scpeters** [comment]
the test seems to fail after writing the test_result xml file, so jenkins doesn't notice that it failed, but it does show a seg-fault in the console log. I've collected a backtrace:

~~~
[       OK ] WideAngleCameraTest.WideAngleCameraBox (5536 ms)
[Thread 0x7ffff26c2700 (LWP 995858) exited]
[----------] 1 test from WideAngleCameraTest (5536 ms total)

[----------] Global test environment tear-down
[Thread 0x7ffff2ec3700 (LWP 995857) exited]
[==========] 1 test from 1 test suite ran. (5536 ms total)
[  PASSED  ] 1 test.
[Thread 0x7ffff36c4700 (LWP 995856) exited]
[Thread 0x7ffff3ec5700 (LWP 995855) exited]
[Thread 0x7ffff46c6700 (LWP 995854) exited]
[Thread 0x7ffff4ec7700 (LWP 995853) exited]
[Thread 0x7ffff56c8700 (LWP 995852) exited]

Thread 1 "INTEGRATION_wid" received signal SIGSEGV, Segmentation fault.
0x00007fffce39dc49 in ?? () from /lib/x86_64-linux-gnu/libGLdispatch.so.0
(gdb) bt
#0  0x00007fffce39dc49 in ?? () from /lib/x86_64-linux-gnu/libGLdispatch.so.0
#1  0x00007fffcc2d0984 in Ogre::GLSL::GLSLProgram::unloadHighLevelImpl() () from /usr/lib/x86_64-linux-gnu/OGRE-1.9.0/RenderSystem_GL.so
#2  0x00007fffce86fd1a in Ogre::HighLevelGpuProgram::unloadHighLevel() () from /lib/x86_64-linux-gnu/libOgreMain.so.1.9.0
#3  0x00007fffce990028 in Ogre::Resource::unload() () from /lib/x86_64-linux-gnu/libOgreMain.so.1.9.0
#4  0x00007fffcc2d08a5 in Ogre::GLSL::GLSLProgram::~GLSLProgram() () from /usr/lib/x86_64-linux-gnu/OGRE-1.9.0/RenderSystem_GL.so
#5  0x00007fffcc2d08bd in Ogre::GLSL::GLSLProgram::~GLSLProgram() () from /usr/lib/x86_64-linux-gnu/OGRE-1.9.0/RenderSystem_GL.so
#6  0x00007fffce619752 in ?? () from /lib/x86_64-linux-gnu/libOgreRTShaderSystem.so.1.9.0
#7  0x00007fffce6156f7 in Ogre::RTShader::ProgramManager::releasePrograms(Ogre::Pass*, Ogre::RTShader::TargetRenderState*) () from /lib/x86_64-linux-gnu/libOgreRTShaderSystem.so.1.9.0
#8  0x00007fffce60288d in Ogre::RTShader::ShaderGenerator::SGTechnique::~SGTechnique() () from /lib/x86_64-linux-gnu/libOgreRTShaderSystem.so.1.9.0
#9  0x00007fffce603751 in Ogre::RTShader::ShaderGenerator::_destroy() () from /lib/x86_64-linux-gnu/libOgreRTShaderSystem.so.1.9.0
#10 0x00007fffce603abe in Ogre::RTShader::ShaderGenerator::destroy() () from /lib/x86_64-linux-gnu/libOgreRTShaderSystem.so.1.9.0
#11 0x00007fffcee0f739 in ignition::rendering::v7::OgreRTShaderSystem::Fini() () from /usr/lib/x86_64-linux-gnu/ign-rendering-7/engine-plugins/libignition-rendering-ogre.so
#12 0x00007fffcee0f902 in ignition::rendering::v7::OgreRTShaderSystem::~OgreRTShaderSystem() () from /usr/lib/x86_64-linux-gnu/ign-rendering-7/engine-plugins/libignition-rendering-ogre.so
#13 0x00007ffff7064a27 in __run_exit_handlers (status=0, listp=0x7ffff7206718 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#14 0x00007ffff7064be0 in __GI_exit (status=<optimized out>) at exit.c:139
#15 0x00007ffff70420ba in __libc_start_main (main=0x5555555657b0 <main(int, char**)>, argc=1, argv=0x7fffffffe288, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe278) at ../csu/libc-start.c:342
#16 0x000055555556586e in _start ()
(gdb)
~~~

could there be a memory problem with one of the shaders?

**Comment by iche033** [comment]
The crash happens after exit. It could be that the OgreRTShaderSystem is trying to destroy its resources after ogre has already been deleted.

**Comment by WilliamLewww** [comment]
> The crash happens after exit. It could be that the OgreRTShaderSystem is trying to destroy its resources after ogre has already been deleted.

I created an integration test for the distortion camera and I am getting the same segfault, but when I delete the following line in ign-rendering, the program exits without a segfault:
https://github.com/ignitionrobotics/ign-rendering/blob/1258280eae5834f5687ad5be78e073e73efb47a2/ogre/src/OgreRTShaderSystem.cc#L156

I think you are correct; the resources (for shaders) are deleted and then Ogre::RtShader::ShaderGenerator tries to delete them again. Removing the **destroy()** might cause memory leaks if there are still allocated resources but I currently don't see any other solutions if it isn't an **ordering of destructors** issue.

**Comment by WilliamLewww** [comment]
The segfault seems to occur with every test that uses OGRE1.

**Review by scpeters — APPROVED** [review]

**Comment by chapulina** [comment]
2 approvals and happy CI, merging!

**Pull request merged**